### ISSUE TITLE
Number / integer

### ISSUE THREAD

**Issue by NicoLaval** [VTL-Community/VTL-Community#44]
User manual, line 1871:

"_A Scalar Value of type number is a rational number of any magnitude and precision, also used as approximation of a real number. On values of type number, the numeric operations are allowed, such as: addition, subtraction, multiplication, division, power, square root and so on. The type integer (positive and negative integer numbers and zero) is a subtype of the type number_"

**Addition**

Operator applied to numbers only with with the following precision, in the Reference manual, line 3032:

"_According to the general rules about data types, the operator can be applied also on sub-types of number, that is the type integer. If the type of both operands is integer then the result has type integer. If one of the operands is of type number, then the other operand is implicitly cast to number and therefore the result has type number._"

I therefore deduce that the following results are correct:

```
a := 1 + 1;
// a = 2 with integer type
b := 1 + 0.1;
// b = 1.1 with number type
```

**nvl**

Operator applied to scalars, not numbers only with with the following additional constraint, in the Reference manual, line 6734:

"_If op1 and op2 are scalar values then they must be of the same type._"

**Is "same type" compatible with the number / integer subtype definition?
It seems to make sense to be able to proceed with the same rule as that defined for addition.**

**What is the good result for the following `b` calculation?**

```
a := nvl(cast(null, integer), 0);
// a = 0 with integer type
b := nvl(cast(null, number), 0);
// b = 0 with number type?
// Throw an exception?
```

**Comment by antonio-olleros** [comment]
Hi Nico, I would say that it is 0 with number type.
The reason is the implicit casting rules. You are comparing a number and an integer, which in principle is not possible. But there is an implicit casting rule from integer to number, so we apply, and the result is that we compare two numbers, and therefore the transformation is correct.

**Comment by NicoLaval** [comment]
Thxs @antonio-olleros.
I agree it's logical to perform this implicit casting.

If this opinion is widely shared, don't you think that this would deserve explicit mention in the reference manual of the future version of the language, in order to clarify this point?

**Comment by vpinna80** [comment]
BUT....
in the new version both INT->NUM and NUM->INT have been made implicit (page 67 of the draft RM). So the result type is ambiguous, unless you either explictly cast one of them or bind the result valudomain to some variable, i.e. in a calc.
This isn't the only one issue with implicit casting, also take a look at https://github.com/sdmx-twg/vtl/issues/383

**Comment by antonio-olleros** [comment]
You are right. In my opinion, we should not be that generous with implicit castings, it should only be allowed when casting does not imply any transformation (from number to integer it implies rounding).
Shall that implicit casting still be kept; we could solve the inconsistency by stating that the implicit casting, when in conflict, should always go to the supertype.

Still, my preferred option is limiting the implicit castings

**Comment by linardian** [comment]
I think that after the cast operator change, this issue could be considered solved; if so, please close it.

**Comment by antonio-olleros** [comment]
Hi @linardian , I think that this issue, per se, can be closed. But as an off-topic, I think that at least @vpinna80 and us agree that we don't like the changes to the cast operator introduced in the published version, where new implicit castings were added in a way that is very difficult (or impossible) to implement. 
On top of that, there is our proposal not to have as implicit a casting from number to string, because most of the times that casting is hiding an error at compile-time, just to get it in run-time... Of course, it may be done on purpose, in which case it should be made explicitly.
Shall we open a new issue to track those two things?

**Comment by linardian** [comment]
Ok, so please close this issue and do open specific issues for each "cell" of the cast operator that could be changed, in order to address the discussions
